### PR TITLE
WPF - Refactor RenderHandler implementations

### PR DESCRIPTION
**Pull request by merceyz**
- Added `AbstractRenderHandler` that takes care of the basics of a `IRenderHandler`
- Refactored the `RenderHandler`s to inherit from `AbstractRenderHandler`
- Implemented the correct and thread safe dispose pattern

### CODE REVIEW

File: CefSharp.Wpf/Rendering/Experimental/ByteArrayWritableBitmapRenderHandler.cs [hunk 1]
```diff
@@ -21,18 +21,14 @@ namespace CefSharp.Wpf.Rendering.Experimental
...
     /// <seealso cref="CefSharp.Wpf.IRenderHandler" />
-    public class ByteArrayWritableBitmapRenderHandler : IRenderHandler
+    public class ByteArrayWritableBitmapRenderHandler : AbstractRenderHandler
     {
-        /// <summary>
-        /// The pixel format
-        /// </summary>
-        private static readonly PixelFormat PixelFormat = PixelFormats.Prgba64;
-        private static int BytesPerPixel = PixelFormat.BitsPerPixel / 8;
+        private static new readonly PixelFormat PixelFormat = PixelFormats.Prgba64;
```
Comment (amaitland) [inline CefSharp.Wpf/Rendering/Experimental/ByteArrayWritableBitmapRenderHandler.cs]: This being `Prgba64` is my mistake, should be `Pbgra32`

File: CefSharp.Wpf/Rendering/AbstractRenderHandler.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                return Interlocked.CompareExchange(ref disposeSignaled, 1, 1) == 1;
+            }
+        }
+
+        /// <summary>
+        /// Releases all resources used by the <see cref="AbstractRenderHandler"/> object
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
```
[the remark this answers is not in the document]
Comment (merceyz) [reply CefSharp.Wpf/Rendering/AbstractRenderHandler.cs]: There is no `Finalizer` implemented so there is no need for it.

[the remark this answers is not in the document]
Comment (amaitland) [reply CefSharp.Wpf/Rendering/AbstractRenderHandler.cs]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implementing-the-dispose-pattern-for-a-base-class still has the call even when there is no `Finalizer`, maybe it's a copy and paste mistake or maybe it has some bearing on references that do implement a `finalizer`.

Happy to be overruled if you have an article reference that says it's completely irrelevant.

[the remark this answers is not in the document]
Comment (merceyz) [reply CefSharp.Wpf/Rendering/AbstractRenderHandler.cs]: You're right, the call should be there in the event any class inheriting from it implements a `Finalizer`.
In this case I knew that doesn't happen but that might, and probably will, change in the future, so I added it.

File: CefSharp.Wpf/Rendering/AbstractRenderHandler.cs [hunk 3]
```diff
@@ -61,6 +61,7 @@ public bool IsDisposed
         public void Dispose()
         {
             Dispose(true);
```
Comment (amaitland) [inline CefSharp.Wpf/Rendering/AbstractRenderHandler.cs]: > If obj does not have a finalizer, the call to the SuppressFinalize method has no effect.

https://docs.microsoft.com/en-us/dotnet/api/system.gc.suppressfinalize?view=netframework-4.7.2

So looks like it's a `NO OP`, not sure why it's in the example.

**Comment by amaitland** [comment]
Will remove the AbstractRenderHandler inheritance from ByteArrayWritableBitmapRenderHandler and make the static properties internal.  

It might be worth creating an abstract class that implants just the Dispose pattern,  will look at doing so if I have time. 

Little more testing and this should be merged in a few days,  will take care of the changes when it comes to merge time.  Thanks :+1:

**Comment by merceyz** [comment]
> It might be worth creating an abstract class that implants just the Dispose pattern, will look at doing so if I have time.

That was what i wanted to do for #2651 but it would only be possible on the OffScreen version.
I also wanted to do it on every object that doesn't have a base class + implements IDisposable but I haven't gotten around to it yet.

File: CefSharp.Wpf/Rendering/AbstractRenderHandler.cs (hunk 3 above)
Comment (merceyz) [reply CefSharp.Wpf/Rendering/AbstractRenderHandler.cs]: It's for when a derived class has a Finalizer

**Comment by amaitland** [comment]
Broadly speaking that sounds fine, feel free to create an issue to track the task. There are quite a few instances where Dispose is called as the result of the unmanaged class/refptr being freed which won't be candidates for a base class.

**Comment by amaitland** [comment]
Follow up commit is https://github.com/cefsharp/CefSharp/commit/4a145bf363cb0854466fd897a5789c844cfd6baf

Changed base class, didn't implement a base disposable class yet, will look at that as a issue of it's own. I guess technically it doesn't need to actually implement `Dispose`, so we might not even bother for `
ByteArrayWritableBitmapRenderHandler`.

**Pull request merged**